Fix max_pool2d with ceil_mode bug

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46558 Fix max_pool2d with ceil_mode bug**

This PR fixes a bug with how pooling output shape was computed.

## BC Breaking Notes
Previously, a bug in the pooling code allowed a sliding window to be entirely off bounds. Now, sliding windows must start inside the input or left padding (not right padding, see https://github.com/pytorch/pytorch/issues/46929) and may only go off-bounds if ceil_mode=True.

## Note
Updating the code to allow the sliding window to start within the right padded region would break many internal assumptions. We may address this in a future PR.

fixes #45357

TODO

- [x] Ensure existing tests are checking for the correct output size

Differential Revision: [D24633372](https://our.internmc.facebook.com/intern/diff/D24633372)

heitorschueroff committed:
Fix max_pool2d with ceil_mode bug

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed. Previously, it did not consider the right padding as "part of the image" but the docs do. Now, in ceil mode it allows a kernel window to go off bounds as long as it starts within the image including both paddings. The boolean condition was also wrong, it should be ceil_mode.

fixes #45357

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed. Previously, it did not consider the right padding as "part of the image" but the docs do. Now, in ceil mode it allows a kernel window to go off bounds as long as it starts within the image including both paddings. The boolean condition was also wrong, it should be ceil_mode.

fixes #45357

TODO

- [ ] Ensure existing tests are checking for the correct output size

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed. 

## BC Breaking Notes
Previously, the pooling code allowed a kernel window to be entirely outside the input and it did not consider right padding as part of the input in the computations. Now, in ceil mode it allows a kernel window to go off bounds as long as it starts within the input including both paddings. If ceil_mode=False, then all kernel windows have to start inside the input, again including both paddings.

fixes #45357

TODO

- [ ] Ensure existing tests are checking for the correct output size

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -1563,6 +1563,17 @@ void _qavg_pool_nhwc_kernel(
           float multiplier = input_scale / output_scale / divide_factor;
           int input_zero_point_m_size = -input_zero_point * size;
 
+          if (size == 0) {
```
Looks like this is causing an ASAN failure and another quantization test failure. What's going on here?

heitorschueroff replied:
The problem was with the AVX2 code when the kernel window is outside the non-padded region. In this case size = 0 so I added a fast-path which also avoids unnecessary loop control logic.

mruberry replied:
Is this necessary? There's no tests in the PR for this quantized code path.

heitorschueroff replied:
removed

mruberry commented on test/test_nn.py:
```diff
@@ -10511,6 +10511,15 @@ def test(nonlinearity, *args, **kwargs):
         test('threshold', 3, 2)
         test('threshold', 3, 2, inplace=True)
 
+    def test_pooling_shape(self, device):
```
Add a comment explaining what this test is for

mruberry commented on test/test_nn.py:
```diff
@@ -10511,6 +10511,15 @@ def test(nonlinearity, *args, **kwargs):
+            t = torch.randn(sizes, device=device)
+            self.assertEqual(torch.nn.functional.max_pool3d(t, *args, **kwargs).shape, expected_out_shape)
+
+        check((1, 1, 3, 4), (1, 1, 6, 7), kernel_size=1, stride=2, padding=0, ceil_mode=True)
```
Add a comment explaining which test case(s) are entirely outside the bounds and what that implies for the shape, vs. those that start with a valid input, vs. those that start in a padded region.

Should one of these test cases have left padding and right padding?

heitorschueroff replied:
I'll add more test cases

mruberry commented on test/test_nn.py:
```diff
@@ -10511,6 +10511,15 @@ def test(nonlinearity, *args, **kwargs):
+    def test_pooling_shape(self, device):
+        def check(expected_out_shape, sizes, *args, **kwargs):
+            t = torch.randn(sizes, device=device)
+            self.assertEqual(torch.nn.functional.max_pool3d(t, *args, **kwargs).shape, expected_out_shape)
```
The issue was for max_pool2d but you're only testing max_pool3d.

heitorschueroff replied:
max_pool2d for quantized tensors calls max_pool3d. Ideally this should test all pooling code, but since they all call the same function to compute output shape, 3D gives me more dimensions to test at once.

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed. 

## BC Breaking Notes
Previously, the pooling code allowed a kernel window to be entirely outside the input and it did not consider right padding as part of the input in the computations. Now, in ceil mode it allows a kernel window to go off bounds as long as it starts within the input including both paddings. If ceil_mode=False, then all kernel windows have to start inside the input, again including both paddings.

fixes #45357

TODO

- [x] Ensure existing tests are checking for the correct output size

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed. 

## BC Breaking Notes
Previously, the pooling code allowed a kernel window to be entirely outside the input and it did not consider right padding as part of the input in the computations. Now, in ceil mode it allows a kernel window to go off bounds as long as it starts within the input including both paddings. If ceil_mode=False, then all kernel windows have to start inside the input, again including both paddings.

fixes #45357

TODO

- [x] Ensure existing tests are checking for the correct output size

[ghstack-poisoned]

mruberry commented on torch/nn/modules/pooling.py:
```diff
@@ -45,6 +45,10 @@ class MaxPool1d(_MaxPoolNd):
     sliding window. This `link`_ has a nice visualization of the pooling parameters.
 
+    Note:
+        When ceil_mode=True, sliding windows are allowed to go off-bounds as long as it starts within the
```
"as long as it..." -> "if they start within the left padding or the input. Windows that would start in the right padded region are ignored."

mruberry commented on test/test_nn.py:
```diff
@@ -10515,6 +10515,22 @@ def test(nonlinearity, *args, **kwargs):
+                        t = torch.randn(sizes[:i + 2], device=device)
+                        self.assertEqual(op(t, *args, **kwargs).shape, expected_out_shape[:i + 2])
+
+        check((1, 1, 3, 3, 4), (1, 1, 5, 6, 7), kernel_size=1, stride=2, padding=0, ceil_mode=True)
```
Would you add the exact test case from the issue, too? The test can cite the issue and test it separately at the end.

mruberry replied:
Would you add the exact test case from the issue, too? The test can cite the issue and test it separately at the end.

mruberry reviewed (approved):
Just two small fixes (doc strings + exact test case from issue). Cool fix!

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed.

## BC Breaking Notes
Previously, a bug in the pooling code allowed a sliding window to be entirely off bounds. Now, sliding windows must start inside the input or left padding (not right padding, see https://github.com/pytorch/pytorch/issues/46929) and may only go off-bounds if ceil_mode=True.

fixes #45357

TODO

- [x] Ensure existing tests are checking for the correct output size

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix max_pool2d with ceil_mode bug"


This PR fixes a bug with how pooling output shape was computed.

## BC Breaking Notes
Previously, a bug in the pooling code allowed a sliding window to be entirely off bounds. Now, sliding windows must start inside the input or left padding (not right padding, see https://github.com/pytorch/pytorch/issues/46929) and may only go off-bounds if ceil_mode=True.

fixes #45357

TODO

- [x] Ensure existing tests are checking for the correct output size

[ghstack-poisoned]
